keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,430 @@
+  - Reason: AwaitingPreemption - Lower priority Pods are being pre-empted,
+  - Reason: FailedNodeCapacity - Scheduler determined Node does not have room,
+* Status: true - Scheduler has completed allocating the resize request,
+  - Reason: Accepted - Pod resize approved by Scheduler.
+
+PodResizeSuccess PodCondition can have the following values:
+* Status: true  - Last resize request was successful,
+  - Reason: Success - Pod and its Containers were resized successfully,
+* Status: false - Last resize request failed,
+  - Reason: FailedNodeCapacity - Kubelet determined Node does not have room,
+  - Reason: RejectedResourceQuota - Resize request exceeds ResourceQuota,
+  - Reason: RejectedQoSClassChange - Resize request changes QoS class,
+  - Reason: Retrying - Controller initiated a retry.
+
+#### Container Restart Policy
+
+To provide some fine-grained control to the user,
+PodSpec.Container.ResourceRequirements is extended with ResizePolicy flag
+for each resource type (CPU, memory) :
+* NoRestart - the default value; resize the Container without restarting it,
+* RestartContainer - restart the Container in-place to apply new resource
+  values (e.g. Java process needs to change its Xmx flag),
+* RestartPod - restart the whole Pod in-place to apply new resource values
+  (e.g. Pod requires its Init Containers to re-run).
+
+By using the ResizePolicy flag, user can mark Containers or Pods as safe
+(or unsafe) for in-place resource update.
+
+This flag is used by Kubelet to determine the actions needed. This flag **may** be
+used by the actors starting the update to decide if the process should be started
+at all (for example VPA might decide to evict Pod with RestartPod policy).
+
+Setting the flag to separately control CPU & memory is due to an observation
+that usually CPU can be added/removed without much problems whereas
+changes to available memory are more probable to require restarts.
+
+If more than one resource type with different policies are updated, then
+RestartPod policy takes precedence over RestartContainer, which in turn takes
+precedence over NoRestart policy.
+
+#### Pod Resize Retry Policy
+
+If resource update fails, say due to lack of space on Node, default behavior
+is to let the initiating actor such as VPA handle the failure. Alternately, a
+Controller can either retry the update, or reschedule the Pod based on policy.
+
+PodSpec is extended with a new flag, PodSpec.RetryPolicy, with possible values:
+* NoRetry - the default value; do nothing, initiating actor handles failure,
+* RetryUpdate - Controller retries resource update in-place,
+* Reschedule - Controller evicts Pod, and creates updated Pod for scheduling.
+
+#### CRI Changes
+
+Kubelet calls UpdateContainerResources CRI API which currently takes
+*runtimeapi.LinuxContainerResources* parameter that works for Docker and Kata,
+but not for Windows. This parameter is changed to *runtimeapi.ContainerResources*,
+that is runtime agnostic.
+
+### Flow Control
```

Comment:
this flow control is really useful.

i would prefer to see if we can build a design that omits the need for the scheduler to be an intermediary.

the pod is already bound, so the desired change in resources should be a relationship between the apiserver and the kubelet, that the scheduler may witness and respond to.

for example:

update desired resources
kubelet sees update and trys to converge to desired state, if it cannot converge, it signals that its not possible.
the scheduler (or some other controller) could see that the kubelet cannot converge, and may choose to do an action to assist the kubelet in converging, but the core interaction must be between the apiserver <-> kubelet without requiring the scheduler to act as an intermediary.  i think that is the only interaction pattern that is backwards compatible.

Reply:
The current proposal has fixed this issue. resolving.